Change deflate internal_state struct size more predictable.

Dead2 opened the pull request:
- Add extra reserved space in deflate internal_state struct for future expansion.
- Make internal_state struct have a static size regardless of what features have been activated.

Internal_state is now always 6040 bytes on Linux/x86-64, and 5952 bytes on Linux/x86-32.

Dead2 committed:
Add extra space in deflate internal_state struct for future expansion.
Also make internal_state struct have a static size regardless of what features have been activated.
Internal_state is now always 6040 bytes on Linux/x86-64, and 5952 bytes on Linux/x86-32.

Dead2 commented:
The crc variables only exist on x86 platforms, since they would never be used on other platforms.
BUT, is that true? Will there at some point be a port of the crc_folding code to ARM or other platforms?

mtl1979 commented:
5916 is not divisible by 8. As most machines nowadays support vector instructions for memory management, it is both safer and more effective to align in 8 byte boundary even on 32-bit operating systems. Even if the next block is allocated using aligned allocation function, it will add implicit padding which is something we should avoid...

Dead2 commented:
From what I can interpret with pahole, I think the struct itself gets 4-byte aligned because it starts with a 32-bit pointer. But the end however is 8-byte aligned due to the int64_t.

But if the calling application does not set its own alloc function, `zng_alloc` will be used, and that alloc will try to align the start of the struct to 64 bytes (to start on a cacheline).
Not sure we can do anything to improve this at this point.

Only alternative I see is requesting a bigger alignment for the struct, but it still gets stuck into an alloc'ed chunk, so I am not sure what that would actually do.

mtl1979 commented:
My point is that the struct size and member offsets should be consistent no matter how the start of the struct is aligned... As 8 byte alignment is largest requirement from the standard, we don't need to worry about larger alignments than that.

nmoinvaz commented on deflate.h:
```diff
@@ -194,7 +194,8 @@ typedef struct internal_state {
     int nice_match; /* Stop searching when current match exceeds this */
 
-#ifdef X86_PCLMULQDQ_CRC
+#if defined(__x86_64__) || defined(_M_X64) || defined(__i386) || defined(_M_IX86)
```
Why not just remove this #define altogether?

Dead2 replied:
Well, if those will never be used by any other arch, then they are just bloat.
If it is likely that we want to use them for other arches as well, then sure, we could keep them around.

nmoinvaz replied:
Ok

nmoinvaz commented on deflate.h:
```diff
@@ -251,17 +252,18 @@ typedef struct internal_state {
     int32_t bi_valid;
     /* Number of valid bits in bi_buf.  All bits above the last valid bit are always zero. */
 
+    int32_t reserved1;
```
Maybe add a comment about why these are _reserved_.

Dead2 replied:
Good idea.

Dead2 replied:
Added

nmoinvaz commented:
Fuzzing ci returns `/usr/bin/ld: compress.lo: relocation R_X86_64_PC32 against symbol `zng_compress2' can not be used when making a shared object; recompile with -fPIC`. I wonder if this is a fluke. Maybe rerun? Hard to think those changes could cause that fuzzing error.

Dead2 commented:
@nmoinvaz Still fails on rerun unfortunately.

mtl1979 commented:
That specific test has been failing for previous pull requests too, but the fail was ignored... It is possible that the failure was introduced by change of clang version... I could possibly trace which was the first pull requests where this specific test failed and compare if clang version is same or if it changed...

There is two issues that I can see... 

1. Signature of compress2() in header and source file doesn't match.
2. The source file containing compress2() is compiled using C compiler, but the library is linked using C++ compiler/linker. This can cause conflict in the symbol tables.

Dead2 commented:
Kind of annoying that the CI Fuzz pipeline only runs on Pull Requests and not Push. Makes it a lot harder to track down when it first failed.

Dead2 commented:
Looks to me like this PR #825 was the first one to fail, and that seems weird since it only changed cmake but the fuzzing seems to run configure..

These were the last successful PR: #824 (and #823 before it)

What makes this more complicated is each PR has its own branch, so the commit history can be different. We really need to run all the CI on Push too.

Between those Fuzzing runs there were changes to the Ubuntu CI image though:
Successful PR #824:
```
  Environment: ubuntu-18.04
  Version: 20201210.0
```
Failing PR #825:
```
  Environment: ubuntu-18.04
  Version: 20210111.1
```

Unfortunately the configure script does not print out compiler version, so I have no idea whether that changed between those two builds.

This does look like some new bug in the Ubuntu packages from the new image. As far as I can tell, no updates are run in the fuzzer CI (except it is attempting to update make), so it'll be running the same versions as exist on the image.

mtl1979 commented:
I think best bet is to run the Fuzzing flow manually on local machine and test if clang version makes difference... Most likely they added some stricter check that wasn't fully tested...

On sidenote, manual patching of LDSHARED should be removed as it is one of the things that can cause weird issues as it changes what libraries are linked against the library being created.

nmoinvaz commented:
CI fuzz only runs on pull request because it won't work with push from a forked repo. It has to do with how Google detects what is the OSS-Fuzz account.

Dead2 commented:
We could perhaps make it conditional of the repo somehow, I remember we had to do some detection like that for the codecov key.

Dead2 commented:
> On sidenote, manual patching of LDSHARED should be removed as it is one of the things that can cause weird issues as it changes what libraries are linked against the library being created.

Do you have a file/line reference for that?

mtl1979 commented:
@Dead2 I tried to find that line from zlib-ng repository, but it seems to come from external repository...

Dead2 commented:
Should we make room here to accomodate the possibly future #828 ?

mtl1979 commented:
@Dead2 I already said the padding should be bigger (especially on 32-bit machines)... But how much bigger, that is just matter of making sure the real size (including explicit and implicit padding) is multiple of 8...

Dead2 commented:
@mtl1979 So I am guessing that you either want me to add an ifdef that adds an extra 4bytes to the end if the architecture is 32bit, or you want me to add a dummy pointer (that adds 4bytes on 32bit and 8bytes on 64bit).

Your answer did not indicate an opinion on the question though, so I will assume that was a yes.

mtl1979 commented:
@Dead2 If last member of the struct is int64_t, total size of the struct should already be multiple of 8, but you claim it isn't... This means the calculation is incorrect in the first place...

What comes to padding at the end of the struct, I think 3x64 bits isn't overkill. If more is required, then we need to know how much more...

Dead2 commented:
The struct was aligned to the required alignment of the first entry, and that was a 32bit pointer, resulting in a 4 byte alignment.
The last entry required 64bit alignment, and the resulting struct ended on a 8 byte alignment.

I can add a alignment hint to the struct so that the compiler assumes 8 bytes alignment for the start of the struct.
The only downside is that this is just a hint for the compiler so it can assume the struct has that alignment, since the memory space the struct gets stuffed into is an alloc, that should still always be safe unless a custom alloc of the application does not align alloc requests on at least 8 bytes. But I think that is an acceptable assumption.

Dead2 commented:
32bit:
```
        /* size: 5952, cachelines: 93, members: 60 */
        /* sum members: 5943, holes: 2, sum holes: 5 */
        /* padding: 4 */
```
64bit:
```
        /* size: 6032, cachelines: 95, members: 60 */
        /* sum members: 6027, holes: 2, sum holes: 5 */
        /* last cacheline: 16 bytes */
```

mtl1979 commented:
@Dead2 I'm not totally sure if it would be better to use exclusively int32_t for padding on 32-bit machine, or just use int64_t exclusively... Mixing 32-bit and 64-bit struct members on 32-bit operating system is prone to inconsistencies...

Dead2 commented:
We already have a uin64_t member in the struct, so we cannot avoid that in any case.
In a previous PR, I took pains to align each member of the struct so they pack as nicely as possible on both 32bit and 64bit while respecting their needed alignments, as a result is we now only have 5bytes of holes on both platforms and both platforms should enjoy an improved cache locality (although that is very hard to optimize for with pointers being in the start of the struct and not at the end).

Dead2 commented:
If we used only uint64_t for reserved, then the removed uint32_t would become an implicit padding on both platforms. Better to keep it explicit IMO.

mtl1979 commented:
@Dead2 I'm just trying to figure out what is the best way to get rid of the 4 byte implicit padding on 32-bit operating systems...

Dead2 commented:
Here you can see how it now gets aligned in respect to cachelines (those can be useful checkpoints that tell you how it is actually aligned in memory).

32bit:
```
struct internal_state {
        zng_stream *               strm;                 /*     0     4 */
        unsigned char *            pending_buf;          /*     4     4 */
        unsigned char *            pending_out;          /*     8     4 */
        uint32_t                   pending_buf_size;     /*    12     4 */
        uint32_t                   pending;              /*    16     4 */
        int                        wrap;                 /*    20     4 */
        uint32_t                   gzindex;              /*    24     4 */
        zng_gz_headerp             gzhead;               /*    28     4 */
        int                        status;               /*    32     4 */
        int                        last_flush;           /*    36     4 */
        int                        reproducible;         /*    40     4 */
        int                        block_open;           /*    44     4 */
        unsigned int               w_size;               /*    48     4 */
        unsigned int               w_bits;               /*    52     4 */
        unsigned int               w_mask;               /*    56     4 */
        unsigned int               lookahead;            /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        unsigned int               high_water;           /*    64     4 */
        unsigned int               window_size;          /*    68     4 */
        unsigned char *            window;               /*    72     4 */
        Pos *                      prev;                 /*    76     4 */
        Pos *                      head;                 /*    80     4 */
        int                        block_start;          /*    84     4 */
        unsigned int               match_length;         /*    88     4 */
        Pos                        prev_match;           /*    92     2 */

        /* XXX 2 bytes hole, try to pack */

        int                        match_available;      /*    96     4 */
        unsigned int               strstart;             /*   100     4 */
        unsigned int               match_start;          /*   104     4 */
        unsigned int               prev_length;          /*   108     4 */
        unsigned int               max_chain_length;     /*   112     4 */
        unsigned int               max_lazy_match;       /*   116     4 */
        int                        level;                /*   120     4 */
        int                        strategy;             /*   124     4 */
        /* --- cacheline 2 boundary (128 bytes) --- */
        unsigned int               good_match;           /*   128     4 */
        int                        nice_match;           /*   132     4 */
        unsigned int               crc0[20];             /*   136    80 */
        /* --- cacheline 3 boundary (192 bytes) was 24 bytes ago --- */
        struct ct_data_s           dyn_ltree[573];       /*   216  2292 */
        /* --- cacheline 39 boundary (2496 bytes) was 12 bytes ago --- */
        struct ct_data_s           dyn_dtree[61];        /*  2508   244 */
        /* --- cacheline 43 boundary (2752 bytes) --- */
        struct ct_data_s           bl_tree[39];          /*  2752   156 */
        /* --- cacheline 45 boundary (2880 bytes) was 28 bytes ago --- */
        struct tree_desc_s         l_desc;               /*  2908    12 */
        struct tree_desc_s         d_desc;               /*  2920    12 */
        struct tree_desc_s         bl_desc;              /*  2932    12 */
        /* --- cacheline 46 boundary (2944 bytes) --- */
        uint16_t                   bl_count[16];         /*  2944    32 */
        int                        heap[573];            /*  2976  2292 */
        /* --- cacheline 82 boundary (5248 bytes) was 20 bytes ago --- */
        int                        heap_len;             /*  5268     4 */
        int                        heap_max;             /*  5272     4 */
        unsigned char              depth[573];           /*  5276   573 */

        /* XXX 3 bytes hole, try to pack */

        /* --- cacheline 91 boundary (5824 bytes) was 28 bytes ago --- */
        unsigned int               lit_bufsize;          /*  5852     4 */
        unsigned char *            sym_buf;              /*  5856     4 */
        unsigned int               sym_next;             /*  5860     4 */
        unsigned int               sym_end;              /*  5864     4 */
        long unsigned int          opt_len;              /*  5868     4 */
        long unsigned int          static_len;           /*  5872     4 */
        unsigned int               matches;              /*  5876     4 */
        unsigned int               insert;               /*  5880     4 */
        long unsigned int          compressed_len;       /*  5884     4 */
        /* --- cacheline 92 boundary (5888 bytes) --- */
        long unsigned int          bits_sent;            /*  5888     4 */
        uint64_t                   bi_buf;               /*  5892     8 */
        int32_t                    bi_valid;             /*  5900     4 */
        int32_t                    reserved1;            /*  5904     4 */
        int64_t                    reserved2[5];         /*  5908    40 */

        /* size: 5952, cachelines: 93, members: 60 */
        /* sum members: 5943, holes: 2, sum holes: 5 */
        /* padding: 4 */
} __attribute__((__aligned__(8)));
```

64bit:
```
struct internal_state {
        zng_stream *               strm;                 /*     0     8 */
        unsigned char *            pending_buf;          /*     8     8 */
        unsigned char *            pending_out;          /*    16     8 */
        uint32_t                   pending_buf_size;     /*    24     4 */
        uint32_t                   pending;              /*    28     4 */
        int                        wrap;                 /*    32     4 */
        uint32_t                   gzindex;              /*    36     4 */
        zng_gz_headerp             gzhead;               /*    40     8 */
        int                        status;               /*    48     4 */
        int                        last_flush;           /*    52     4 */
        int                        reproducible;         /*    56     4 */
        int                        block_open;           /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        unsigned int               w_size;               /*    64     4 */
        unsigned int               w_bits;               /*    68     4 */
        unsigned int               w_mask;               /*    72     4 */
        unsigned int               lookahead;            /*    76     4 */
        unsigned int               high_water;           /*    80     4 */
        unsigned int               window_size;          /*    84     4 */
        unsigned char *            window;               /*    88     8 */
        Pos *                      prev;                 /*    96     8 */
        Pos *                      head;                 /*   104     8 */
        int                        block_start;          /*   112     4 */
        unsigned int               match_length;         /*   116     4 */
        Pos                        prev_match;           /*   120     2 */

        /* XXX 2 bytes hole, try to pack */

        int                        match_available;      /*   124     4 */
        /* --- cacheline 2 boundary (128 bytes) --- */
        unsigned int               strstart;             /*   128     4 */
        unsigned int               match_start;          /*   132     4 */
        unsigned int               prev_length;          /*   136     4 */
        unsigned int               max_chain_length;     /*   140     4 */
        unsigned int               max_lazy_match;       /*   144     4 */
        int                        level;                /*   148     4 */
        int                        strategy;             /*   152     4 */
        unsigned int               good_match;           /*   156     4 */
        int                        nice_match;           /*   160     4 */
        unsigned int               crc0[20];             /*   164    80 */
        /* --- cacheline 3 boundary (192 bytes) was 52 bytes ago --- */
        struct ct_data_s           dyn_ltree[573];       /*   244  2292 */
        /* --- cacheline 39 boundary (2496 bytes) was 40 bytes ago --- */
        struct ct_data_s           dyn_dtree[61];        /*  2536   244 */
        /* --- cacheline 43 boundary (2752 bytes) was 28 bytes ago --- */
        struct ct_data_s           bl_tree[39];          /*  2780   156 */
        /* --- cacheline 45 boundary (2880 bytes) was 56 bytes ago --- */
        struct tree_desc_s         l_desc;               /*  2936    24 */
        /* --- cacheline 46 boundary (2944 bytes) was 16 bytes ago --- */
        struct tree_desc_s         d_desc;               /*  2960    24 */
        struct tree_desc_s         bl_desc;              /*  2984    24 */
        /* --- cacheline 47 boundary (3008 bytes) --- */
        uint16_t                   bl_count[16];         /*  3008    32 */
        int                        heap[573];            /*  3040  2292 */
        /* --- cacheline 83 boundary (5312 bytes) was 20 bytes ago --- */
        int                        heap_len;             /*  5332     4 */
        int                        heap_max;             /*  5336     4 */
        unsigned char              depth[573];           /*  5340   573 */

        /* XXX 3 bytes hole, try to pack */

        /* --- cacheline 92 boundary (5888 bytes) was 28 bytes ago --- */
        unsigned int               lit_bufsize;          /*  5916     4 */
        unsigned char *            sym_buf;              /*  5920     8 */
        unsigned int               sym_next;             /*  5928     4 */
        unsigned int               sym_end;              /*  5932     4 */
        long unsigned int          opt_len;              /*  5936     8 */
        long unsigned int          static_len;           /*  5944     8 */
        /* --- cacheline 93 boundary (5952 bytes) --- */
        unsigned int               matches;              /*  5952     4 */
        unsigned int               insert;               /*  5956     4 */
        long unsigned int          compressed_len;       /*  5960     8 */
        long unsigned int          bits_sent;            /*  5968     8 */
        uint64_t                   bi_buf;               /*  5976     8 */
        int32_t                    bi_valid;             /*  5984     4 */
        int32_t                    reserved1;            /*  5988     4 */
        int64_t                    reserved2[5];         /*  5992    40 */

        /* size: 6032, cachelines: 95, members: 60 */
        /* sum members: 6027, holes: 2, sum holes: 5 */
        /* last cacheline: 16 bytes */
} __attribute__((__aligned__(8)));
```

mtl1979 commented:
@Dead2 It's clear that it does align int64_t/uint64_t on 4 byte boundary as if it is array of two 32-bit variables... This causes the implicit padding... Only way to fix it is to add int32_t as last member of the struct or change reserved1 to be int32_t[2]  only on 32-bit machines .

Dead2 commented:
Added a pointer before the uint64_t to improve its alignment, this also got rid of the implicit padding at the end of the struct.
Converted all reserved bytes to int32_t of same total size for simplicity.

32bit:
```
struct internal_state {
        zng_stream *               strm;                 /*     0     4 */
        unsigned char *            pending_buf;          /*     4     4 */
        unsigned char *            pending_out;          /*     8     4 */
        uint32_t                   pending_buf_size;     /*    12     4 */
        uint32_t                   pending;              /*    16     4 */
        int                        wrap;                 /*    20     4 */
        uint32_t                   gzindex;              /*    24     4 */
        zng_gz_headerp             gzhead;               /*    28     4 */
        int                        status;               /*    32     4 */
        int                        last_flush;           /*    36     4 */
        int                        reproducible;         /*    40     4 */
        int                        block_open;           /*    44     4 */
        unsigned int               w_size;               /*    48     4 */
        unsigned int               w_bits;               /*    52     4 */
        unsigned int               w_mask;               /*    56     4 */
        unsigned int               lookahead;            /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        unsigned int               high_water;           /*    64     4 */
        unsigned int               window_size;          /*    68     4 */
        unsigned char *            window;               /*    72     4 */
        Pos *                      prev;                 /*    76     4 */
        Pos *                      head;                 /*    80     4 */
        int                        block_start;          /*    84     4 */
        unsigned int               match_length;         /*    88     4 */
        Pos                        prev_match;           /*    92     2 */

        /* XXX 2 bytes hole, try to pack */

        int                        match_available;      /*    96     4 */
        unsigned int               strstart;             /*   100     4 */
        unsigned int               match_start;          /*   104     4 */
        unsigned int               prev_length;          /*   108     4 */
        unsigned int               max_chain_length;     /*   112     4 */
        unsigned int               max_lazy_match;       /*   116     4 */
        int                        level;                /*   120     4 */
        int                        strategy;             /*   124     4 */
        /* --- cacheline 2 boundary (128 bytes) --- */
        unsigned int               good_match;           /*   128     4 */
        int                        nice_match;           /*   132     4 */
        unsigned int               crc0[20];             /*   136    80 */
        /* --- cacheline 3 boundary (192 bytes) was 24 bytes ago --- */
        struct ct_data_s           dyn_ltree[573];       /*   216  2292 */
        /* --- cacheline 39 boundary (2496 bytes) was 12 bytes ago --- */
        struct ct_data_s           dyn_dtree[61];        /*  2508   244 */
        /* --- cacheline 43 boundary (2752 bytes) --- */
        struct ct_data_s           bl_tree[39];          /*  2752   156 */
        /* --- cacheline 45 boundary (2880 bytes) was 28 bytes ago --- */
        struct tree_desc_s         l_desc;               /*  2908    12 */
        struct tree_desc_s         d_desc;               /*  2920    12 */
        struct tree_desc_s         bl_desc;              /*  2932    12 */
        /* --- cacheline 46 boundary (2944 bytes) --- */
        uint16_t                   bl_count[16];         /*  2944    32 */
        int                        heap[573];            /*  2976  2292 */
        /* --- cacheline 82 boundary (5248 bytes) was 20 bytes ago --- */
        int                        heap_len;             /*  5268     4 */
        int                        heap_max;             /*  5272     4 */
        unsigned char              depth[573];           /*  5276   573 */

        /* XXX 3 bytes hole, try to pack */

        /* --- cacheline 91 boundary (5824 bytes) was 28 bytes ago --- */
        unsigned int               lit_bufsize;          /*  5852     4 */
        unsigned char *            sym_buf;              /*  5856     4 */
        unsigned int               sym_next;             /*  5860     4 */
        unsigned int               sym_end;              /*  5864     4 */
        long unsigned int          opt_len;              /*  5868     4 */
        long unsigned int          static_len;           /*  5872     4 */
        unsigned int               matches;              /*  5876     4 */
        unsigned int               insert;               /*  5880     4 */
        long unsigned int          compressed_len;       /*  5884     4 */
        /* --- cacheline 92 boundary (5888 bytes) --- */
        long unsigned int          bits_sent;            /*  5888     4 */
        char *                     reserved_p;           /*  5892     4 */
        uint64_t                   bi_buf;               /*  5896     8 */
        int32_t                    bi_valid;             /*  5904     4 */
        int32_t                    reserved_1[11];       /*  5908    44 */

        /* size: 5952, cachelines: 93, members: 60 */
        /* sum members: 5947, holes: 2, sum holes: 5 */
} __attribute__((__aligned__(8)));
```

64bit:
```
struct internal_state {
        zng_stream *               strm;                 /*     0     8 */
        unsigned char *            pending_buf;          /*     8     8 */
        unsigned char *            pending_out;          /*    16     8 */
        uint32_t                   pending_buf_size;     /*    24     4 */
        uint32_t                   pending;              /*    28     4 */
        int                        wrap;                 /*    32     4 */
        uint32_t                   gzindex;              /*    36     4 */
        zng_gz_headerp             gzhead;               /*    40     8 */
        int                        status;               /*    48     4 */
        int                        last_flush;           /*    52     4 */
        int                        reproducible;         /*    56     4 */
        int                        block_open;           /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        unsigned int               w_size;               /*    64     4 */
        unsigned int               w_bits;               /*    68     4 */
        unsigned int               w_mask;               /*    72     4 */
        unsigned int               lookahead;            /*    76     4 */
        unsigned int               high_water;           /*    80     4 */
        unsigned int               window_size;          /*    84     4 */
        unsigned char *            window;               /*    88     8 */
        Pos *                      prev;                 /*    96     8 */
        Pos *                      head;                 /*   104     8 */
        int                        block_start;          /*   112     4 */
        unsigned int               match_length;         /*   116     4 */
        Pos                        prev_match;           /*   120     2 */

        /* XXX 2 bytes hole, try to pack */

        int                        match_available;      /*   124     4 */
        /* --- cacheline 2 boundary (128 bytes) --- */
        unsigned int               strstart;             /*   128     4 */
        unsigned int               match_start;          /*   132     4 */
        unsigned int               prev_length;          /*   136     4 */
        unsigned int               max_chain_length;     /*   140     4 */
        unsigned int               max_lazy_match;       /*   144     4 */
        int                        level;                /*   148     4 */
        int                        strategy;             /*   152     4 */
        unsigned int               good_match;           /*   156     4 */
        int                        nice_match;           /*   160     4 */
        unsigned int               crc0[20];             /*   164    80 */
        /* --- cacheline 3 boundary (192 bytes) was 52 bytes ago --- */
        struct ct_data_s           dyn_ltree[573];       /*   244  2292 */
        /* --- cacheline 39 boundary (2496 bytes) was 40 bytes ago --- */
        struct ct_data_s           dyn_dtree[61];        /*  2536   244 */
        /* --- cacheline 43 boundary (2752 bytes) was 28 bytes ago --- */
        struct ct_data_s           bl_tree[39];          /*  2780   156 */
        /* --- cacheline 45 boundary (2880 bytes) was 56 bytes ago --- */
        struct tree_desc_s         l_desc;               /*  2936    24 */
        /* --- cacheline 46 boundary (2944 bytes) was 16 bytes ago --- */
        struct tree_desc_s         d_desc;               /*  2960    24 */
        struct tree_desc_s         bl_desc;              /*  2984    24 */
        /* --- cacheline 47 boundary (3008 bytes) --- */
        uint16_t                   bl_count[16];         /*  3008    32 */
        int                        heap[573];            /*  3040  2292 */
        /* --- cacheline 83 boundary (5312 bytes) was 20 bytes ago --- */
        int                        heap_len;             /*  5332     4 */
        int                        heap_max;             /*  5336     4 */
        unsigned char              depth[573];           /*  5340   573 */

        /* XXX 3 bytes hole, try to pack */

        /* --- cacheline 92 boundary (5888 bytes) was 28 bytes ago --- */
        unsigned int               lit_bufsize;          /*  5916     4 */
        unsigned char *            sym_buf;              /*  5920     8 */
        unsigned int               sym_next;             /*  5928     4 */
        unsigned int               sym_end;              /*  5932     4 */
        long unsigned int          opt_len;              /*  5936     8 */
        long unsigned int          static_len;           /*  5944     8 */
        /* --- cacheline 93 boundary (5952 bytes) --- */
        unsigned int               matches;              /*  5952     4 */
        unsigned int               insert;               /*  5956     4 */
        long unsigned int          compressed_len;       /*  5960     8 */
        long unsigned int          bits_sent;            /*  5968     8 */
        char *                     reserved_p;           /*  5976     8 */
        uint64_t                   bi_buf;               /*  5984     8 */
        int32_t                    bi_valid;             /*  5992     4 */
        int32_t                    reserved[11];         /*  5996    44 */

        /* size: 6040, cachelines: 95, members: 60 */
        /* sum members: 6035, holes: 2, sum holes: 5 */
        /* last cacheline: 24 bytes */
} __attribute__((__aligned__(8)));
```

mtl1979 commented:
Looks better now... I don't think we can fit anything useful in the 2 and 3 byte holes, so adding dummy struct members to fill them makes little sense...

nmoinvaz commented on deflate.h:
```diff
@@ -251,18 +252,22 @@ typedef struct internal_state {
     unsigned long bits_sent;      /* bit length of compressed data sent mod 2^32 */
-#endif
+
+    /* Reserved for future use */
```
Maybe just add a comment to this to indicate it is here for alignment purposes? `Reserved for future use and alignment`. And same with the comment one below it.

Dead2 replied:
Fixed.

nmoinvaz reviewed (approved):
LGTM

nmoinvaz commented:
@Dead2 how did you manually generate those struct alignment data or do you have a tool that does it for you?

Dead2 commented:
@nmoinvaz I use the `pahole` program on linux, AFAIK it only works for ELF binaries and it requires debuginfo.
